.github/workflows/ci.yml

```diff
@@ -13,159 +13,103 @@ concurrency:
-            c_compiler: clang
-            cpp_compiler: clang++
+            container: null
+          - os: ubuntu-22.04
+            toolchain: gcc
+            c_compiler: gcc
+            cpp_compiler: g++
+            container: dockcross/manylinux_2_28-x64:latest
         exclude:
-          - os: windows-latest
-            c_compiler: clang
-          - os: windows-latest
-            c_compiler: cl
-            build_type: Debug
-          - os: ubuntu-latest
-            c_compiler: cl
-          - os: macos-latest
-            c_compiler: cl
-
-    name: CMake ${{ matrix.os }} ${{ matrix.build_type }} (${{ matrix.c_compiler }})
+            - os: macos-14
+              toolchain: msvc
+            - os: macos-14
+              toolchain: gcc
+            - os: ubuntu-22.04
+              toolchain: msvc
+            - os: ubuntu-22.04
+              toolchain: clang
+            - os: windows-2022
+              toolchain: clang
+            - os: windows-2022
+              toolchain: gcc
+
+    name: CMake-${{ matrix.os }}-${{ matrix.build_type }}-${{ matrix.toolchain }}
     runs-on: ${{ matrix.os }}
+    container: ${{ matrix.container }}
+    env:
+      SCCACHE_GHA_ENABLED: "true"
 
     steps:
-    - uses: actions/checkout@v3
+    - uses: actions/checkout@v4
 
-    - name: Set reusable strings
-      # Turn repeated input strings (such as the build output directory) into step outputs. These step outputs can be used throughout the workflow file.
-      id: strings
-      shell: bash
-      run: |
-        echo "build-output-dir=${{ github.workspace }}/out/build" >> "$GITHUB_OUTPUT"
-
-    - name: Install Ubuntu build dependencies
-      if: matrix.os == 'ubuntu-latest'
-      run: |
-        sudo apt update
-        sudo apt install libgl-dev libx11-xcb-dev libxrandr-dev libxinerama-dev libxcursor-dev libxi-dev libxext-dev wayland-protocols libwayland-dev libxkbcommon-dev
-
-    - name: Set up Python
-      uses: actions/setup-python@v4
-      with:
-        python-version: 3.11
-
-    - name: Fetch dependencies
-      run: python ./tools/fetch_dawn_dependencies.py --use-test-deps --shallow
-
-    - name: Set up Ninja
-      uses: seanmiddleditch/gha-setup-ninja@master
-
-    - name: Set up Visual Studio environment
-      if: matrix.os == 'windows-latest'
-      uses: seanmiddleditch/gha-setup-vsdevenv@master
+    - name: Set up dependencies on linux
+      if: matrix.container == 'dockcross/manylinux_2_28-x64:latest'
+      run: >
+        dnf install -y mesa-libGL-devel libxcb libxcb-devel libX11-xcb libXcursor-devel libXrandr-devel libXinerama-devel libXi-devel libXext-devel libxkbcommon libxkbcommon-devel libxkbcommon-x11-devel mesa-vulkan-drivers wayland-protocols-devel wayland-devel
 
     - name: Set up sccache
       uses: mozilla-actions/sccache-action@v0.0.3
 
     - name: Configure CMake
       run: >
-        cmake -G Ninja -B ${{ steps.strings.outputs.build-output-dir }}
+        cmake 
```

Comment:
Any reason to not use Ninja?

Reply:
I wanted to minimize number of external dependencies in the workflow file.

1. for mac and linux, the default generator is already Ninja. 
2. for windows, two reasons to use VS:
  a. the VS generator (default) lets cmake initialize the VS development environment variables like path to cl.exe, windows sdk and so on.
  b. the PDB locking issue on debug builds which was documented in the previous commit message does not manifest with VS generator, so that's another pro.

Reply:
Thanks for the explanation!